### PR TITLE
Development guide

### LINKED ISSUE #213: Create developer setup guide
This could be on the wiki or in a doc, but we need some kind of documentation on how to get going.

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #213

This should address the need for a dev guide so new developers can hack on the bridge. It's currently drafted and needs proofreading.

### CODE REVIEW

File: docs/development_guide.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+called `slack-bridge-env`:
+
+```bash
+mkdir ~/slack-bridge-env
+cd ~/slack-bridge-env
+```
+
+### Setting up Synapse
+
+Largely to setup Synapse, you can follow https://hub.docker.com/r/matrixdotorg/synapse and just ensure your data directory points to `~/slack-bridge-env/synapse`.
```
Comment (turt2live) [inline docs/development_guide.md]: There's a change of voice/tone here which is a little jarring: the previous section uses a voice where the guide is the user ("we", "our", etc) but here it changes to be instructional ("do this", "you", etc). Personally I prefer the more fuzzy feeling of the guide saying "we", "us", "our", and such.

File: docs/development_guide.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    "defaultCountryCode": "GB",
+    "showLabsSettings": false,
+    "default_federate": true,
+    "default_theme": "light"
+}
+```
+
+Finally, run `docker run -v /home/will/git/scalar-env/riot-config.json:/app/config.json -p 8080:80 vectorim/riot-web` to start your Riot instance. You should be able to register a new user on your local synapse instance through Riot.
+
+### Setting up the bridge postgres
```
Comment (turt2live) [inline docs/development_guide.md]:
```suggestion
### Setting up the bridge with PostgreSQL
```

Comment (turt2live) [reply docs/development_guide.md]: also might be relevant to link to digital ocean's managed databases thing. It's a fairly inexpensive option for people doing lots of development

File: docs/development_guide.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+```
+
+You should also create a new database in preparation for using the bridge
+
+```bash
+sudo apt install postgresql-client # Ensure you have psql installed
+psql -h localhost -U postgres -W postgres
+```
+
+And then on the postgres shell
```
Comment (turt2live) [inline docs/development_guide.md]:
```suggestion
And then on the postgres shell:
```

File: docs/development_guide.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    "default_federate": true,
+    "default_theme": "light"
+}
+```
+
+Finally, run `docker run -v /home/will/git/scalar-env/riot-config.json:/app/config.json -p 8080:80 vectorim/riot-web` to start your Riot instance. You should be able to register a new user on your local synapse instance through Riot.
+
+### Setting up the bridge postgres
+
+Setting up the postgres instance for the bridge is also quite easy.
```
Comment (turt2live) [inline docs/development_guide.md]: Repetition. Also the things should be numbered so people can say "I need help with step 3". Worth noting that guides over ~10 steps are considered tutorials and usually take a blog post format.

File: docs/development_guide.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+Make sure that you copy the generated registration file to `~/slack-bridge-env/synapse` and add an entry for it in the `homeserver.yaml` before starting the synapse container.
+
+## Making changes
+
+Whenever you want to make changes to the codebase, you must:
+
+```bash
+git fetch matrix-org
+git checkout matrix-org/develop
+git checkout -b yourfeaturname
```
Comment (turt2live) [inline docs/development_guide.md]:
```suggestion
git checkout -b yourfeaturename
```

File: docs/development_guide.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+npm i
+```
+
+You should always work within the `src` directory. It is helpful to have a code editor setup with linting enabled so you can see mistakes as you work. Always remember to run `npm run build` before commiting or testing so that you know the latest changes work.
```
Comment (turt2live) [inline docs/development_guide.md]: line wrapping is inconsistent

File: docs/development_guide.md (hunk 1 above)
Comment (benparsons) [reply docs/development_guide.md]: The docker commands in the hub.docker.com docs need to be changed. I recommend including them explicitly here with the needed changes, for example:

```
docker run -it --rm \
    -v ~/slack-bridge-env/synapse:/data \
    -e SYNAPSE_SERVER_NAME=localhost \
    -e SYNAPSE_REPORT_STATS=no \
    matrixdotorg/synapse:latest generate

docker run -d --name synapse \
    -v ~/slack-bridge-env/synapse:/data \
    -p 8008:8008 \
    matrixdotorg/synapse:latest
```

File: docs/development_guide.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+Setting up the postgres instance for the bridge is also quite easy.
+
+```bash
+docker run -d --name slackpg -p 59999:5432 -e POSTGRES_PASSWORD=pass postgres
+```
+
+You should also create a new database in preparation for using the bridge
+
+```bash
+sudo apt install postgresql-client # Ensure you have psql installed
```
Comment (benparsons) [inline docs/development_guide.md]: can we separate this out and include the following line, useful for dev on a mac:

```
brew install libpg
```

I think this is worth being explicit about because `brew install postgres` installs the entire server on the host, which is not what's needed here.

File: docs/development_guide.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+git checkout matrix-org/develop # Always base your changes off matrix-org/develop
+npm i
+npm run build
+```
+
+The steps above should make sure you are running the latest development version
+and have built the typescript. To ensure that it's all working, you can run
+`npm test` which will run the unit and integration tests. 
+
+You can follow the instructions in the [README](../README.md) to generate and update the registration file, as well as creating a testing Slack app and workspace. You should enable RTM support in the config, as Slack will not be able to push events to your local bridge.
```
Comment (benparsons) [inline docs/development_guide.md]: Two options:

either the README needs to be more structured so you can point to numbers or headings that need to be followed.

OR

Extract the relevant parts of the README and repeat here

File: docs/development_guide.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+```bash
+mkdir ~/slack-bridge-env
+cd ~/slack-bridge-env
+```
+
+### Setting up Synapse
+
+Largely to setup Synapse, you can follow https://hub.docker.com/r/matrixdotorg/synapse and just ensure your data directory points to `~/slack-bridge-env/synapse`.  
+You will want to make sure your server name is something that routes to your local box. I tend to use the devbox hostname, but `localhost` is also sufficent.  
+You should enable registration in the `homeserver.yaml` to create a testing user.
```
Comment (benparsons) [inline docs/development_guide.md]: Can be explicit:

Find and uncomments the `enable_registration` field, set it to `true`:

```
enable_registration: true
```

File: docs/development_guide.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    "disable_3pid_login": false,
+    "brand": "Riot",
+    "defaultCountryCode": "GB",
+    "showLabsSettings": false,
+    "default_federate": true,
+    "default_theme": "light"
+}
+```
+
+Finally, run `docker run -v /home/will/git/scalar-env/riot-config.json:/app/config.json -p 8080:80 vectorim/riot-web`
```
Comment (benparsons) [inline docs/development_guide.md]: docker run -d --name riot -v ~/slack-bridge-env/riot-config.json:/app/config.json -p 8080:80 vectorim/riot-web

File: docs/development_guide.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+You can setup PostgreSQL in docker.
+
+```bash
+docker run -d --name slackpg -p 59999:5432 -e POSTGRES_PASSWORD=pass postgres
+```
+
+You should also create a new database in preparation for using the bridge
+
+```bash
+psql -h localhost -U postgres -W postgres
```
Comment (benparsons) [inline docs/development_guide.md]: Running this at this point returns

```
psql: could not connect to server: Connection refused
	Is the server running on host "localhost" (::1) and accepting
	TCP/IP connections on port 5432?
could not connect to server: Connection refused
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 5432?
```

Is there a missing step here?

File: README.md [hunk 12]
```diff
@@ -3,6 +3,14 @@ A Matrix &lt;--> Slack bridge
 
 This bridge allows you to connect Slack channels to Matrix rooms.
 
+Development
+-----------
+
+[![#slack:half-shot.uk](https://img.shields.io/matrix/slack:half-shot.uk.svg?server_fqdn=matrix.half-shot.uk&label=%23slack:half-shot.uk&logo=matrix)](https://matrix.to/#/#slack:half-shot.uk)
+
+If you want to help out, please give our helpful [Development Guide](./docs/development_guide.md)
+a read. It covers all you need to know to get hacking on the bridge.
```
Comment (benparsons) [inline README.md]: Can we be more explicit about the Setup steps and how they relate to getting a dev environment working? Eg port selection could be specific, the config file "you may wish to copy and edit as appropriate" - this can be really prescriptive for dev environment rather than deployment.

File: docs/development_guide.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+-----------------
+
+This guide is for those that want to help hack on the Slack bridge, and need some instruction on how
+to set it up. This is **not** a guide for setting up an instance for production usage.
+
+For information on how to contribute issues and or pull requests, please read [CONTRIBUTING](../CONTRIBUTING.md).
+
+## Setting up your environment
+
+This section explains how to setup Synapse, Riot and the Slack bridge for local development.
```
[the remark this answers is not in the document]
Comment (turt2live) [reply docs/development_guide.md]: I seriously disagree with that decision. For one, I'm not reviewing the same blobs of text over and over, and I'm sure not expecting that anyone goes around to update all the install guides for all the projects. 

Pointing to an existing guide has a two major benefits as far as I'm concerned:

* The project retains control over which installation method is preferred. The upstream projects (namely Riot and Synapse) are likely to want to change their preferred installation methods, particularly Riot in the nearish future. By nature of the relationship between the bridge team and the auxiliary projects, the auxiliary projects would need to inform the bridge team that their docs are now outdated, leading to a lot of suddenly outdated documentation.
* The steps are in a single canonical location. There's several bridges and other projects which would need this same guide: by rewriting the same words over and over, we're duplicating information. Pointing to the project's own guides means we don't have to repeat it.

In short: please let's not make more work for everyone.

[the remark this answers is not in the document]
Comment (benparsons) [reply docs/development_guide.md]: My problem with keeping the setup instructions in their own home is that the specific requirements for this purpose (setup to create a dev environment for hacking on the Slack bridge) are not made clear by the instructions for those projects. The way it is right now - with very minimal steps needed for each case, plus a link to the canonical sources is much more effective. See this example: https://github.com/matrix-org/matrix-appservice-slack/pull/215/files#diff-c570356b15a684fa012ea55e4347e51bR73, which I found very readable.

What's currently not readable is having the dev setup guide link to the README and DATASTORES pages, which are much more general pages that go beyond the scope of what's needed for a dev environment setup.

It *may* be more work to maintain these duplicated blobs (I'm not sure how much more work it would be), but right now it's too complex to get started hacking on this project, so more streamlined docs are neeeded.

[the remark this answers is not in the document]
Comment (turt2live) [reply docs/development_guide.md]: Setting up a Synapse server for development or for public is largely the same, barring a few config changes. I seriously don't see a problem with saying "Install Synapse as per [their instructions] then add the appservice's registration file to the config. If you're not sure how to do that, [click here]". Both links would (and should) be to the Synapse repository.

File: .buildkite/pipeline.yml [hunk 14]
```diff
@@ -37,7 +37,7 @@ steps:
     branches: "!master !develop !release-*"
     command:
       - "python3 -m pip install towncrier"
-      - "python3 -m towncrier.check"
+      - "python3 -m towncrier.check --compare-with=origin/develop"
```
Comment (auscompgeek) [inline .buildkite/pipeline.yml]: This change seems like it should be separate from this PR.

Comment (Half-Shot) [reply .buildkite/pipeline.yml]: Yep, we should factor that out.

Comment (Half-Shot) [reply .buildkite/pipeline.yml]: https://github.com/matrix-org/matrix-appservice-slack/pull/258

**Comment by Half-Shot** [comment]
Hey there people on the tail end of 2025, closing this PR as development on the Slack bridge has been wound down now that the foundation is no longer running an instance. I'll work with the foundation on how we might want to communicate the status of this repository, but for now I'll close out any PRs that will not be landing.
